src/caffe/layers/batch_norm_layer.cu

```diff
@@ -6,166 +6,205 @@
 
 namespace caffe {
 
+// multicast x[c] into y[.,c,...]
+template <typename Dtype>
+void BatchNormLayer<Dtype>::multicast_gpu(int N, int C, int S,
+    const Dtype *x, Dtype *y ) {
+  Blob<Dtype> temp_NC;
+  vector<int> temp_size;
+  temp_size.push_back(N*C);
+  temp_NC.Reshape(temp_size);
+  caffe_gpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N, C, 1,
+      1., ones_N_.gpu_data(), x, 0., temp_NC.mutable_gpu_data());
+  caffe_gpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N*C, S , 1,
+      1., temp_NC.gpu_data(), ones_HW_.gpu_data(), 0., y);
+}
+
+// y[c] = sum x(.,c,...)
+template <typename Dtype>
+void BatchNormLayer<Dtype>::compute_sum_per_channel_gpu(int N, int C, int S,
+    const Dtype *x, Dtype *y ) {
+  // assume that x.shape(1)==C
+  Blob<Dtype> temp_NC;
+  vector<int> temp_size;
+  temp_size.push_back(N*C);
+  temp_NC.Reshape(temp_size);
+  caffe_gpu_gemv<Dtype>(CblasNoTrans, N * C, S, 1., x, ones_HW_.gpu_data(),
+      0., temp_NC.mutable_gpu_data());
+  caffe_gpu_gemv<Dtype>(CblasTrans, N, C, 1., temp_NC.gpu_data(),
+      ones_N_.gpu_data(), 0., y);
+}
+
+// y[c] = mean x(.,c,...)
+template <typename Dtype>
+void BatchNormLayer<Dtype>::compute_mean_per_channel_gpu(int N, int C, int S,
+    const Dtype *x, Dtype *y ) {
+  Dtype F = 1.0 / (N*S);
+  compute_sum_per_channel_gpu(N, C, S, x, y);
+  caffe_gpu_scale(C, F, y, y);
+}
+
 template <typename Dtype>
 void BatchNormLayer<Dtype>::Forward_gpu(const vector<Blob<Dtype>*>& bottom,
     const vector<Blob<Dtype>*>& top) {
+  int N = bottom[0]->shape(0);
+  int C = channels_;
+  int S = bottom[0]->count(0) / (N*C);
+  int top_size = top[0]->count();
+
   const Dtype* bottom_data = bottom[0]->gpu_data();
   Dtype* top_data = top[0]->mutable_gpu_data();
-  int num = bottom[0]->shape(0);
-  int spatial_dim = bottom[0]->count()/(channels_*bottom[0]->shape(0));
-
-  if (bottom[0] != top[0]) {
-    caffe_copy(bottom[0]->count(), bottom_data, top_data);
-  }
 
+//  if (bottom[0] != top[0]) {
```

Comment:
Right. I left it as a reminder to test BN in-place with potential overwrite ( e.g. con. layer --> BN --> relu)